core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/dimension/legacy/HighCardDictDimensionIndexCodec.java

```diff
@@ -17,71 +17,117 @@
+        int[] lengthArray = new int[data.length];
+        int max = Integer.MIN_VALUE;
+        int min = Integer.MAX_VALUE;
+        int currentDataLength;
+        int size = 0;
+        for (int i = 0; i < lengthArray.length; i++) {
+          currentDataLength = data[i].length;
+          lengthArray[i] = currentDataLength;
+          size += currentDataLength;
+          if (max < currentDataLength) {
+            max = currentDataLength;
+          }
+          if (min > currentDataLength) {
+            min = currentDataLength;
+          }
         }
-        byte[] flattened = ByteUtil.flatten(indexStorage.getDataPage());
         Compressor compressor = CompressorFactory.getInstance().getCompressor(
             input.getColumnCompressorName());
+        pageIndexGenerator =
+              new BinaryPageIndexGenerator(data, isSort, lengthArray);
+        // free memory
+        selectedDataType = fitLongMinMax(max, min);
+        byte[][] dataPage = pageIndexGenerator.getDataPage();
+        ByteBuffer byteBuffer;
+        if (DataTypes.BYTE == selectedDataType) {
+          byteBuffer = ByteBuffer.allocate(lengthArray.length + size);
+          for (int i = 0; i < lengthArray.length; i++) {
+            byteBuffer.put((byte) lengthArray[i]);
+            byteBuffer.put(dataPage[i]);
+          }
+        } else if (DataTypes.SHORT == selectedDataType) {
+          byteBuffer = ByteBuffer.allocate((lengthArray.length * 2) + size);
+          for (int i = 0; i < lengthArray.length; i++) {
+            byteBuffer.putShort((short) lengthArray[i]);
+            byteBuffer.put(dataPage[i]);
+          }
+        } else if (DataTypes.SHORT_INT == selectedDataType) {
+          byteBuffer = ByteBuffer.allocate((lengthArray.length * 3) + size);
+          for (int i = 0; i < lengthArray.length; i++) {
+            byteBuffer.put(ByteUtil.to3Bytes(lengthArray[i]));
+            byteBuffer.put(dataPage[i]);
+          }
+        } else {
+          byteBuffer = ByteBuffer.allocate((lengthArray.length * 4) + size);
+          for (int i = 0; i < lengthArray.length; i++) {
+            byteBuffer.putInt(lengthArray[i]);
+            byteBuffer.put(dataPage[i]);
+          }
+        }
+        byteBuffer.rewind();
+        byte[] flattened = byteBuffer.array();
         super.compressedDataPage = compressor.compressByte(flattened);
-        super.indexStorage = indexStorage;
+        super.pageIndexGenerator = pageIndexGenerator;
       }
 
-      @Override
-      protected List<Encoding> getEncodingList() {
-        List<Encoding> encodings = new ArrayList<>();
-        if (isVarcharType) {
-          encodings.add(Encoding.DIRECT_COMPRESS_VARCHAR);
-        } else if (indexStorage.getRowIdPageLengthInBytes() > 0) {
-          encodings.add(Encoding.INVERTED_INDEX);
-        }
-        if (indexStorage.getDataRlePageLengthInBytes() > 0) {
-          encodings.add(Encoding.RLE);
+      private DataType fitLongMinMax(int max, int min) {
```

Comment:
Better move this method to encoding strategy

Reply:
ok